dotnet/src/SemanticKernel/Memory/Collections/ScoredValue.cs

```diff
@@ -18,11 +18,18 @@ public ScoredValue(T item, double score)
     }
 
     public T Value { get; }
-    public Score Score { get; }
+    public double Score { get; }
 
     public int CompareTo(ScoredValue<T> other)
```

Comment:
What is the intended usage of ScoredValue? I'm not sure what to make of the `Value` playing a role in its `CompareTo`.

Reply:
@stephentoub The main purpose of `ScoredValue` structure is to be used in `TopNCollection` (it's `MinHeap`, which keeps only top `N` entries by score). We use this collection to find top `N` similar records from DB based on `Score`. So, the actual value doesn't matter, and we need to sort entries only by score. Primary scenario in SDK is when `T` is `MemoryRecord` object.

In `ScoredValue` structure we have following methods and operators to implement - `CompareTo`, `Equals`, `GetHashCode`, and `<`, `>`, `<=`, `>=`, `==`, `!=`. If we want to keep all method and operator results consistent - we will have to do the comparison in all places using the same set of properties:
- In case we use both `Value` and `Score` in all places - all results will be consistent when `T Value` implements `IComparable` interface.
- In case we use just `Score` in all places - comparison operators will be consistent, while I'm not sure how `Equals` and `GetHashCode` will serve their purpose (for example in `Dictionary`, even though it's not the main purpose of `ScoredValue`, but it's still `public` and accessible by users). 

Currently in `main` branch we use just `Score` for comparison, while both `Value` and `Score` for equality, which creates inconsistency between `<=`, `>=` and `==` operators, that's why GH issue and this PR were created.

I'm wondering, what would be the right solution here, because it looks like we have to sacrifice some functionality to make it work consistent for SDK use-case only.

Reply:
> The main purpose of ScoredValue structure is to be used in TopNCollection

Do we need TopNCollection and ScoredValue to public?  Could we just rely on OrderBy(...).Take(...)?

Reply:
> Do we need TopNCollection and ScoredValue to public? Could external consumers just rely on OrderBy(...).Take(...)?

Good questions. As far as I can see they are used in a couple of memory connectors. Also, we still call `TopNCollection.SortByScore` method manually, which performs in-place sorting. I'm not sure about performance advantages here, taking into account that we shouldn't operate with big amount of records in runtime, ideally the sorting should be performed on memory DB side. 

@shawncal Do you have any thoughts regarding `OrderBy(...).Take(...)` instead of `TopNCollection`?

Reply:
The only real advantage to TopNCollection is that, as values are added, it keeps only the top (scoring) values, discarding others. This makes for very fast additions, as there are only N items to compare to at any given time. Most other collections would hold onto every item, then sort all of those items, then remove most of those items, putting potentially a lot more memory pressure on these searches.

That said, I would very much like to remove TopNCollection and ScoredValue from the public API surface. We could just copy these classes into the few connectors that use them, or we could use those linq methods in the connectors and use these classes in our own internal implementation only.